alacritty/Cargo.toml

```diff
@@ -30,6 +30,7 @@ copypasta = { version = "0.7.0", default-features = false }
 libc = "0.2"
 unicode-width = "0.1"
 bitflags = "1"
+memoffset = "0.5"
```

Comment:
See my comment about `offset_of!` we're already handling attrib pointers with it, so I think you can try what we're doing.

Reply:
I have trouble figuring out how to amend that macro to be valid and available for both mod.rs and solidrect.rs

Moreover, I think that doing it the way I do it is better: it more clearly states that this particular field is supposed to be this attribute.

Reply:
Yeah, but it's 3 build deps and memoffset testing their crate in a way we compute address. The thing is that it works quite good only for `repr(C)` structs. The real solution to what we're doing is when rust provides a **reliable** way to take an offset into struct field.

Example of the patch that gets rid of `memoffset`.

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 1064cb2..1eddd10 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -38,7 +38,6 @@ dependencies = [
  "image",
  "libc",
  "log",
- "memoffset",
  "notify",
  "objc",
  "parking_lot",
@@ -1141,15 +1140,6 @@ dependencies = [
  "winapi 0.3.9",
 ]
 
-[[package]]
-name = "memoffset"
-version = "0.5.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "043175f069eda7b85febe4a74abbaeff828d9f8b448515d3151a14a3542811aa"
-dependencies = [
- "autocfg",
-]
-
 [[package]]
 name = "miniz_oxide"
 version = "0.3.7"
diff --git a/alacritty/Cargo.toml b/alacritty/Cargo.toml
index 9069b93..081ea4d 100644
--- a/alacritty/Cargo.toml
+++ b/alacritty/Cargo.toml
@@ -30,7 +30,6 @@ copypasta = { version = "0.7.0", default-features = false }
 libc = "0.2"
 unicode-width = "0.1"
 bitflags = "1"
-memoffset = "0.5"
 
 [build-dependencies]
 gl_generator = "0.14.0"
diff --git a/alacritty/src/renderer/rectrenderer.rs b/alacritty/src/renderer/rectrenderer.rs
index f2c72af..eb5f9d4 100644
--- a/alacritty/src/renderer/rectrenderer.rs
+++ b/alacritty/src/renderer/rectrenderer.rs
@@ -1,8 +1,6 @@
 use std::mem::size_of;
 use std::ptr;
 
-use memoffset::offset_of;
-
 use alacritty_terminal::term::SizeInfo;
 
 use crate::gl;
@@ -104,27 +102,34 @@ impl RectRenderer {
             // VBO binding is not part ot VAO, but attributes are.
             gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
 
+            let mut size = 0;
+            let mut index = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $gl_normalized:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        $gl_normalized,
+                        size_of::<Vertex>() as i32,
+                        size as *const _,
+                    );
+                    gl::EnableVertexAttribArray(index);
+
+                    #[allow(unused_assignments)]
+                    {
+                        size += $count * size_of::<$type>();
+                        index += 1;
+                    }
+                };
+            }
+
             // Position.
-            gl::VertexAttribPointer(
-                0,
-                2,
-                gl::FLOAT,
-                gl::FALSE,
-                (size_of::<Vertex>()) as _,
-                ptr::null(),
-            );
-            gl::EnableVertexAttribArray(0);
+            add_attr!(2, gl::FLOAT, gl::FALSE, f32);
 
             // Color.
-            gl::VertexAttribPointer(
-                1,
-                4,
-                gl::UNSIGNED_BYTE,
-                gl::TRUE,
-                (size_of::<Vertex>()) as _,
-                offset_of!(Vertex, color) as *const _,
-            );
-            gl::EnableVertexAttribArray(1);
+            add_attr!(4, gl::UNSIGNED_BYTE, gl::TRUE, u8);
 
             // Reset buffer bindings.
             gl::BindVertexArray(0);

```

Can go futher and refactor the macro in quad render to share this macro in rect render.

Reply:
Yeah I tried making it common and ran into a wall as I still don't understand how to do macros in rust properly (and it modules `use` model drives me nuts).